### PR TITLE
Improve frame docstring and data validation

### PR DESCRIPTION
**Pull request by smoh**
This is my attempt to address #7911.

> 1. That the differentials must be at least 1/time and there is no unitless time (although there is unitless length) should be noted somewhere, and the error message printed when they are not can be improved?

The components of differential data is validated to be time-derivative of corresponding component of representation data.

> 2. Current docstring templating gives wrong information about initializing Galactic frame with the cartesian representation.

I think the ideal would be to do a sort of self-reference for the class's empty instance (because you have to do e.g., `Galactic().representation_info` to access the property), but I'm not sure if that is a good idea, so I insert a caveat and next possible course of action.

### CODE REVIEW

File: astropy/coordinates/baseframe.py [hunk 1]
```diff
@@ -341,14 +341,18 @@ def __new__(cls, reprname, framename, defaultunit='recommended'):
         sets the expected input representation class, thereby changing the
         expected keyword arguments for the data passed in. For example, passing
         ``representation_type='cartesian'`` will make the classes expect
-        position data with cartesian names, i.e. ``x, y, z`` in most cases.
+        position data with cartesian names, i.e. ``x, y, z`` in most cases
+        unless overriden via `frame_specific_representation_info`.
+        Check out `<this frame>().representation_info`.
```
Comment (Juanlu001) [inline astropy/coordinates/baseframe.py]: This makes the docs fail:

```
/home/circleci/project/build/lib.linux-x86_64-3.6/astropy/coordinates/builtin_frames/altaz.py:docstring of astropy.coordinates.AltAz:40:py:obj reference target not found: <this frame>().representation_info
```

(https://circleci.com/gh/astropy/astropy/38707)

File: astropy/coordinates/baseframe.py [hunk 2]
```diff
@@ -341,14 +341,18 @@ def __new__(cls, reprname, framename, defaultunit='recommended'):
...
     differential_type : `~astropy.coordinates.BaseDifferential` subclass, str, dict, optional
         A differential class or dictionary of differential classes (currently
         only a velocity differential with key 's' is supported). This sets the
         expected input differential class, thereby changing the expected keyword
         arguments of the data passed in. For example, passing
         ``differential_type='cartesian'`` will make the classes expect velocity
-        data with the argument names ``v_x, v_y, v_z``.
+        data with the argument names ``v_x, v_y, v_z`` unless overriden via
+        `frame_specific_representation_info`.
+        Check out `<this frame>().representation_info`.
```
Comment (Juanlu001) [inline astropy/coordinates/baseframe.py]: See above

File: astropy/coordinates/tests/test_frames_with_velocity.py [hunk 3]
```diff
@@ -307,3 +307,16 @@ def test_negative_distance():
...
+    """Check that the differential data given has compatible units
+    with the time-derivative of representation data"""
+    with pytest.raises(ValueError) as excinfo:
+        c = ICRS(
+            x=1, y=2, z=3,
+            v_x=1, v_y=2, v_z=3,
+            representation_type=r.CartesianRepresentation,
+            differential_type=r.CartesianDifferential)
+    assert "data units are not compatible with" in str(excinfo.value)
+
```
Comment (Juanlu001) [inline astropy/coordinates/tests/test_frames_with_velocity.py]:
```
astropy/coordinates/tests/test_frames_with_velocity.py:322:1: W391 blank line at end of file
```

**Comment by smoh** [comment]
Markdown habits.. fixed.

**Comment by Juanlu001** [comment]
The 32bit errors are https://github.com/astropy/astropy/issues/9059. I clicked "rebuild" because there was a timeout (but unfortunately started rebuilding everything, sorry :crossed_fingers:)

**Comment by bsipocz** [comment]
> The 32bit errors are #9059. I clicked "rebuild" because there was a timeout (but unfortunately started rebuilding everything, sorry 🤞)

there is a dropdown that offers "restart job with ssh" to restart just the one failing build.

**Comment by Juanlu001** [comment]
AppVeyor, on the other hand, displays a super weird error:

```
Traceback (most recent call last):
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1062, in __call__
    return func(*args, **kwargs)
  File "C:\conda\lib\site-packages\conda\cli\main.py", line 84, in _main
    exit_code = do_call(args, p)
  File "C:\conda\lib\site-packages\conda\cli\conda_argparse.py", line 80, in do_call
    module = import_module(relative_mod, __name__.rsplit('.', 1)[0])
  File "C:\conda\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\conda\lib\site-packages\conda\cli\main_install.py", line 8, in <module>
    from .install import install
  File "C:\conda\lib\site-packages\conda\cli\install.py", line 19, in <module>
    from ..core.index import calculate_channel_urls, get_index
  File "C:\conda\lib\site-packages\conda\core\index.py", line 9, in <module>
    from .package_cache_data import PackageCacheData
  File "C:\conda\lib\site-packages\conda\core\package_cache_data.py", line 15, in <module>
    from conda_package_handling.api import InvalidArchiveError
  File "C:\conda\lib\site-packages\conda_package_handling\api.py", line 3, in <module>
    from libarchive.exception import ArchiveError as _LibarchiveArchiveError
  File "C:\conda\lib\site-packages\libarchive\__init__.py", line 1, in <module>
    from .entry import ArchiveEntry
  File "C:\conda\lib\site-packages\libarchive\entry.py", line 6, in <module>
    from . import ffi
  File "C:\conda\lib\site-packages\libarchive\ffi.py", line 48, in <module>
    libarchive = ctypes.cdll.LoadLibrary(libarchive_path)
  File "C:\conda\lib\ctypes\__init__.py", line 434, in LoadLibrary
    return self._dlltype(name)
  File "C:\conda\lib\ctypes\__init__.py", line 356, in __init__
    self._handle = _dlopen(self._name, mode)
TypeError: LoadLibrary() argument 1 must be str, not None
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\conda\Scripts\conda-script.py", line 12, in <module>
    sys.exit(main())
  File "C:\conda\lib\site-packages\conda\cli\main.py", line 150, in main
    return conda_exception_handler(_main, *args, **kwargs)
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1354, in conda_exception_handler
    return_value = exception_handler(func, *args, **kwargs)
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1065, in __call__
    return self.handle_exception(exc_val, exc_tb)
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1109, in handle_exception
    return self.handle_unexpected_exception(exc_val, exc_tb)
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1120, in handle_unexpected_exception
    self.print_unexpected_error_report(error_report)
  File "C:\conda\lib\site-packages\conda\exceptions.py", line 1190, in print_unexpected_error_report
    from .cli.main_info import get_env_vars_str, get_main_info_str
  File "C:\conda\lib\site-packages\conda\cli\main_info.py", line 19, in <module>
    from ..core.index import _supplement_index_with_system
  File "C:\conda\lib\site-packages\conda\core\index.py", line 9, in <module>
    from .package_cache_data import PackageCacheData
  File "C:\conda\lib\site-packages\conda\core\package_cache_data.py", line 15, in <module>
    from conda_package_handling.api import InvalidArchiveError
  File "C:\conda\lib\site-packages\conda_package_handling\api.py", line 7, in <module>
    from .tarball import CondaTarBZ2 as _CondaTarBZ2
  File "C:\conda\lib\site-packages\conda_package_handling\tarball.py", line 7, in <module>
    import libarchive
  File "C:\conda\lib\site-packages\libarchive\__init__.py", line 1, in <module>
    from .entry import ArchiveEntry
  File "C:\conda\lib\site-packages\libarchive\entry.py", line 6, in <module>
    from . import ffi
  File "C:\conda\lib\site-packages\libarchive\ffi.py", line 48, in <module>
    libarchive = ctypes.cdll.LoadLibrary(libarchive_path)
  File "C:\conda\lib\ctypes\__init__.py", line 434, in LoadLibrary
    return self._dlltype(name)
  File "C:\conda\lib\ctypes\__init__.py", line 356, in __init__
    self._handle = _dlopen(self._name, mode)
TypeError: LoadLibrary() argument 1 must be str, not None
ERROR: the last command returned the following exit code: 1
```

**Comment by bsipocz** [comment]
Ignore appveyor, we stopped using it and switched to travis windows on master already. To make sure all is good, this PR can be rebased.

Also @Juanlu001, you need to set the milestone and weather it needs a changelog or not.

File: astropy/coordinates/tests/test_frames_with_velocity.py [hunk 4]
```diff
@@ -307,3 +307,15 @@ def test_negative_distance():
...
+def test_velocity_units():
+    """Check that the differential data given has compatible units
+    with the time-derivative of representation data"""
+    with pytest.raises(ValueError) as excinfo:
+        c = ICRS(
+            x=1, y=2, z=3,
+            v_x=1, v_y=2, v_z=3,
+            representation_type=r.CartesianRepresentation,
+            differential_type=r.CartesianDifferential)
```
Comment (Juanlu001) [inline astropy/coordinates/tests/test_frames_with_velocity.py]: For the record: the error with current `master` is

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/juanlu/Development/Astropy/astropy/astropy/coordinates/baseframe.py", line 575, in __init__
    {'s': differential_data})
  File "/home/juanlu/Development/Astropy/astropy/astropy/coordinates/representation.py", line 704, in with_differentials
    new_rep._validate_differentials(differentials))
  File "/home/juanlu/Development/Astropy/astropy/astropy/coordinates/representation.py", line 517, in _validate_differentials
    .format(repr(diff), expected_key, key))
ValueError: For differential object '<CartesianDifferential (d_x, d_y, d_z) [dimensionless]
    (1., 2., 3.)>', expected unit key = '' but received key = 's'
```

**Comment by eteq** [comment]
Hmm.  This is a tricky one w.r.t. milestone. It’s sort of a behavior change in that existing code might start raising a `ValueError` where it didn’t before. Very strictly speaking that's an API change.  But I don't think we formally documented the expected error or whether this even is an error, so at other times we've considered these things to be "bug fixes" in that it better reflects *intended* behavior.

My thinking is that that means we should split the difference and only backport to 3.2.x, not all the way to 2.0.x.  But maybe I'm being too cautious if we're sure this only "re-labels" the existing error rather than creating a new one in some corner cases?

**Comment by smoh** [comment]
I know nothing about backporting this to how far back but: I do not think that this change introduces any new `ValueError`. As @Juanlu001  commented, there already is a ValueError that would have been raised anyway down the line, and in my opinion, that ValueError does not make sense. It says, for unitless-time-differential of unitless lengths, you expect an empty string (`''`) as the key but got `'s'` instead. There is seemingly no wrong doing from the user side according to this, except that there is no support for unitless-time and internally the code always expects a quantity compatible with `[units of length] / [unitful time]`.

Also, wouldn't something else in the coordinates module tests have failed if this extra ValueError is not compatible with any usage?

**Comment by eteq** [comment]
Ohh, gotcha, @smoh, I misunderstood that comment and thought it raises a `ValueError` in a *different* spot.  I guess then it's all a question of if there are any corner cases where this still yields a new `ValueError`, then.  Since this is already a bit of a corner case, I'd guess no, in which case back-porting further back is fine.

**Comment by eteq** [comment]
Having said, that, I noticed some peculiar behavior when trying to compare your test to the example I suggested adding in #9110 .  Specifically, the following:
```
>>> i = ICRS(x=1,y=2,z=3, representation_type=CartesianRepresentation)
>>> i.realize_frame(i.data.with_differentials(CartesianDifferential(1,2,3))
ValueError: Multiple differentials are associated with the representation object passed in to the frame initializer. Only a single velocity differential is supported. Got: {'': <CartesianDifferential (d_x, d_y, d_z) [dimensionless]
    (1., 2., 3.)>}
```

Now this is fine in that they all are still giving `ValueError`'s.  But this would appear to be the exact same case you're catching in the `__init__`.  The confusing thing to me is that it looks to me like this should *still* hit your case since `realize_frame()` calls `_replicate()` which eventually calls the class initializer... It's not worth holding up this PR very long to fix this, but if there's an obvious way to adjust it so that we get the same error in this case as in your test, that would be a bonus.

**Comment by eteq** [comment]
To not delay any further I pushed up a changelog update consistent with @adrn's suggestion, and with that I think this is ready when CI passes.

**Comment by smoh** [comment]
I'll just leave it up to the maintainers. Upon examining the example @eteq posted (a while ago), I found that this pull request does not fix everything but I have not had time to come up with a cleaner solution. But perhaps we can fix things one at a time and that's ok.

**Comment by bsipocz** [comment]
Needs a rebase to pick up CI changes.

**Comment by bsipocz** [comment]
Thanks @smoh!
